src/query/sql/src/evaluator/block_operator.rs

```diff
@@ -88,201 +85,117 @@ impl BlockOperator {
-    /// | [1,2]   |      0 |
-    /// | [3,4,5] |      1 |
-    /// +---------+--------+
-    /// ```
-    ///
-    /// After `unnest(arr)`, each row of `number` will be replicated by the length of `arr` at the same row.
-    /// The result will be:
-    ///
-    /// ```
-    /// +-------------+--------+
-    /// | unnest(arr) | number |
-    /// +-------------+--------+
-    /// | 1           |      0 |
-    /// | 2           |      0 |
-    /// | 3           |      1 |
-    /// | 4           |      1 |
-    /// | 5           |      1 |
-    /// +-------------+--------+
-    /// ```
-    ///
-    /// If the argument of `unnest` is a scalar, like:
-    ///
-    /// ```sql
-    /// select unnest([1,2,3]), number from numbers(2);
-    /// ```
-    ///
-    /// The array scalar `[1,2,3]` will be replicated first (See the logic in `BlockOperator::execute`).
-    fn fit_unnest(
-        input: DataBlock,
-        unnest_columns: Vec<ArrayColumn<AnyType>>,
-    ) -> Result<DataBlock> {
-        if unnest_columns.is_empty() {
-            return Ok(input);
-        }
+                let mut result_data_blocks = Vec::with_capacity(input.num_rows());
+                for i in 0..input.num_rows() {
+                    let mut row = Vec::with_capacity(input.num_rows());
 
-        let (unnest_columns, unnest_offsets) = Self::unify_unnest_columns(unnest_columns);
-        let num_rows = *unnest_offsets.last().unwrap();
+                    // Get the max number of rows of all result sets.
+                    let mut max_num_rows = 0;
+                    result.iter().for_each(|srf_results| {
+                        let (_, result_set_rows) = srf_results.get(i).unwrap();
+                        if *result_set_rows > max_num_rows {
+                            max_num_rows = *result_set_rows;
+                        }
+                    });
 
-        // Convert unnest_offsets to take indices.
-        let mut take_indices = Vec::with_capacity(num_rows);
-        for (i, offset) in unnest_offsets.windows(2).enumerate() {
-            take_indices.extend(vec![i as u64; offset[1] - offset[0]]);
-        }
+                    if max_num_rows == 0 && !result_data_blocks.is_empty() {
+                        // Skip current row
+                        continue;
+                    }
 
-        let mut cols = Vec::with_capacity(input.num_columns());
-        let meta = input.get_meta().cloned();
+                    for entry in input.columns() {
+                        // Take the i-th row of input data block and add it to the row.
+                        let mut builder =
+                            ColumnBuilder::with_capacity(&entry.data_type, max_num_rows);
+                        let scalar_ref = entry.value.index(i).unwrap();
+                        (0..max_num_rows).for_each(|_| {
+                            builder.push(scalar_ref.clone());
+                        });
+                        row.push(BlockEntry {
```

Comment:
These codes will replicate much useless data (for example: the args of the SRF, the columns need to be removed after projection). How about replace the SRF columns in place.

Reply:
The arguments of SRF won't be put into the `DataBlock`, only the columns that need to be pass through `ProjectSet` will be replicated.